hibernate-core/src/main/java/org/hibernate/stat/internal/StatsConcurrentHashMap.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.stat.internal;
+
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+/**
+ * Decorates a ConcurrentHashMap implementation to make sure the methods are being
+ * used correctly for the purpose of Hibernate's statistics.
+ * In particular, we do like the semantics of ConcurrentHashMap#computeIfAbsent
+ * but not how it performs.
+ * See also:
+ *  - http://jsr166-concurrency.10961.n7.nabble.com/Re-ConcurrentHashMap-computeIfAbsent-td11687.html
+ *  - https://hibernate.atlassian.net/browse/HHH-13527
+ *
+ * @author Sanne Grinovero
+ */
+final class StatsConcurrentHashMap<V> {
+
+	private final ConcurrentHashMap<String,V> map = new ConcurrentHashMap();
+
+	public void clear() {
+		map.clear();
+	}
+
+	/**
+	 * This method is inherently racy and expensive. Only use on non-hot paths, and
+	 * only to get a recent snapshot.
+	 * @return all keys in string form.
+	 */
+	public String[] keysAsArray() {
+		//We don't allocate an array directly, nor use streams, as its a concurrent map which we don't want to lock:
+		ArrayList<String> collector = new ArrayList<>( map.size() );
+		for ( Map.Entry<String,V> e : map.entrySet() ) {
+			collector.add( e.getKey() );
```

Comment:
Why use ```map.entrySet()```? If ```map.keySet()``` is used instead, then there would be no need to call ```e.getKey()```.

Javadoc for ```ConcurrentHashMap``` doesn't seem to mention why ```entrySet()``` would preferred over ```keySet()```.

Reply:
keySet() creates a fully copy of the key collection into a set: it's less efficient.

Reply:
Looks like what I said is no longer true since java8, I could change it back to keySet and have the same impact.